File: javascript/trapping-rain-water-ii.js
/**
 * @param {number[][]} heightMap
 * @return {number}
 */
function trapRainWater(heightMap) {
    const queue = new PriorityQueue({ compare: ([h1], [h2]) => h1 - h2 });
    const visited = Array.from(heightMap, (row) => new Array(row.length).fill(false));
    // adding top and bottom edges
    for (const row of [0, heightMap.length - 1]) {
        for (let col = 0; col < heightMap[row].length; col++) {
            queue.enqueue([heightMap[row][col], row, col]);
            visited[row][col] = true;
        }
    }
    // adding left and right edges
    for (let row = heightMap.length - 2; row >= 1; row--) {
        queue.enqueue([heightMap[row][0], row, 0]);
        visited[row][0] = true;

        const lastCol = heightMap[row].length - 1;
        queue.enqueue([heightMap[row][lastCol], row, lastCol]);
        visited[row][lastCol] = true;
    }

    let maxHeight = -1;
    let totalVolume = 0;
    while (!queue.isEmpty()) {
        const [height, r, c] = queue.dequeue();
        maxHeight = Math.max(maxHeight, height);
        totalVolume += maxHeight - height;

        const neighbors = [[r, c + 1], [r, c - 1], [r + 1, c], [r - 1, c]]
            .filter(([y, _]) => 0 <= y && y < heightMap.length)
            .filter(([y, x]) => 0 <= x && x < heightMap[y].length)
            .filter(([y, x]) => !visited[y][x]);
        for (const [y, x] of neighbors) {
            queue.enqueue([heightMap[y][x], y, x]);
            visited[y][x] = true;
        }
    }
    return totalVolume;
}

File: python3/trapping-rain-water-ii.py
import heapq


class Solution:
    def trapRainWater(self, heightMap: list[list[int]]) -> int:
        queue = []

        rows = len(heightMap)
        cols = len(heightMap[0])
        visited = [[False] * len(row) for row in heightMap]

        for r, row in enumerate(heightMap):
            for c, cell in enumerate(row):
                if r in (0, rows - 1) or c in (0, cols - 1):
                    heapq.heappush(queue, (cell, r, c))
                    visited[r][c] = True
        res = 0
        max_height = 0

        while len(queue) > 0:
            height, r, c = heapq.heappop(queue)
            
            max_height = max(max_height, height)
            res += max_height - height

            for nr, nc in [(r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)]:
                if nr < 0 or len(heightMap) <= nr:
                    continue
                if nc < 0 or len(heightMap[nr]) <= nc:
                    continue
                if visited[nr][nc]:
                    continue
                heapq.heappush(queue, (heightMap[nr][nc], nr, nc))
                visited[nr][nc] = True
        return res
